docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,181 @@
+
+##### Positive
+
+- Can prioritize urgent transactions.
+- More deterministic than the previous pure auction system.
+
+##### Negative
+
+- Unpredictable.
+- Overpricing the tip price in theory TODO I guess we don't obverse this in real world on ethereum ?
+
+#### Multi-tier pricing system
+
+[Recent IOHK blog](https://iohk.io/en/blog/posts/2021/11/26/network-traffic-and-tiered-pricing/) described a multi-tier gas pricing system. User can specify the tier of service needed for the transaction. The higher the tier, the higher priority, and higher gas price. Within the same tier, transactions pay the same gas price. It avoids the first price auction while provides prioritized service for urgent transactions.
+
+The original proposal reserve block segments for different tiers, we can't do that without modify tendermint, but it should be good enough to just specify higher priority for higher tier transactions. 
+
+The lowest tier can use constant gas price, the higher tiers adjust gas price based on block load at different speed.
+
+##### Positive
+
+- Predictable at the lowest/default tier.
+- Can prioritize urgent transactions in higher tiers.
+- Deterministic at all tiers, no price auction at all.
+
+##### Negative
+
+- TODO
+
+## Decision
+
+After comparing the trade-offs of different designs, we think the adapted multi-tier system is most suitable for cosmos-sdk's case.
+
+We can adopt EIP-1559's approach to adjust higher tier's gas prices based on block load.
+
+### Implementation
+
+We can make each tier's gas price strategy fully configurable in protocol parameters, while providing a sensible default one.
+
+Pseudocode in python-like syntax:
+
+```python
+interface TieredTx:
+  def tier(self) -> int:
+    pass
+
+def tx_tier(tx):
+    if isinstance(tx, TieredTx):
+      return tx.tier()
+    else:
+      # default tier for custom transactions
+      return 0
+
+class TierParams:
+  'gas price strategy parameters of one tier'
+  priority: int           # priority in tendermint mempool
+  initial_gas_price: Coin
+  parent_gas_target: int
+  change_speed: Decimal   # 0 means don't adjust for block load.
+
+class Params:
+    'protocol parameters'
+    tiers: List[TierParams]
+
+class State:
+    'consensus state'
+    # total gas used in last block, None when it's the first block
+    parent_gas_used: Optional[int]
+    # gas prices of last block for all tiers
+    gas_prices: List[Coin]
+
+def begin_block():
+    'Adjust gas prices'
+    for i, tier in enumerate(Params.tiers):
+        if State.parent_gas_used is None:
+            # initialized gas price for the first block
+	          State.gas_prices[i] = tier.initial_gas_price
+        else:
+            gas_used_change_rate = (State.parent_gas_used - tier.parent_gas_target) // tier.parent_gas_target
+            State.gas_prices[i] *= 1 + gas_used_change_rate * tier.change_speed
```

Comment:
I think there is something wrong, according to this formula the gas price will be always growing.

Reply:
If `State.parent_gas_used < tier.parent_gas_target`, `gas_used_change_rate` will be negative and `State.gas_prices[i]` will drop.

Reply:
ah, makes sense.